model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java

```diff
@@ -0,0 +1,875 @@
+                .collect(Collectors.collectingAndThen(
+                        Collectors.toList(),
+                        list -> {
+                            if (list.size() == 0) {
+                                return null;
+                            } else if (list.size() != 1) {
+                                throw new IllegalStateException("More results found for identityProvider=" + socialLink.getIdentityProvider() +
+                                        ", userId=" + socialLink.getUserId() + ", results=" + list);
+                            }
+
+                            return entityToAdapterFunc(realm).apply(list.get(0));
+                        }));
+    }
+
+    @Override
+    public void addConsent(RealmModel realm, String userId, UserConsentModel consent) {
+        LOG.tracef("addConsent(%s, %s, %s)%s", realm, userId, consent, getShortStackTrace());
+
+        UserConsentEntity consentEntity = UserConsentEntity.fromModel(consent);
+        getRegisteredEntityById(realm, userId).ifPresent(userEntity -> userEntity.addUserConsent(consentEntity));
+    }
+
+    @Override
+    public UserConsentModel getConsentByClient(RealmModel realm, String userId, String clientInternalId) {
+        LOG.tracef("getConsentByClient(%s, %s, %s)%s", realm, userId, clientInternalId, getShortStackTrace());
+        return getEntityById(realm, userId)
+                .map(userEntity -> userEntity.getUserConsent(clientInternalId))
+                .map(consent -> UserConsentEntity.toModel(realm, consent))
+                .orElse(null);
+    }
+
+    @Override
+    public Stream<UserConsentModel> getConsentsStream(RealmModel realm, String userId) {
+        LOG.tracef("getConsentByClientStream(%s, %s)%s", realm, userId, getShortStackTrace());
+        return getRegisteredEntityById(realm, userId)
+                .map(AbstractUserEntity::getUserConsents)
+                .orElse(Stream.empty())
+                .map(consent -> UserConsentEntity.toModel(realm, consent));
+    }
+
+    @Override
+    public void updateConsent(RealmModel realm, String userId, UserConsentModel consent) {
+        LOG.tracef("updateConsent(%s, %s, %s)%s", realm, userId, consent, getShortStackTrace());
+
+        MapUserEntity user = getEntityByIdOrThrow(realm, userId);
+        UserConsentEntity userConsentEntity = user.getUserConsent(consent.getClient().getId());
+        if (userConsentEntity == null) {
+            throw new ModelException("Consent or not found for client [" + consent.getClient().getId() + "] and user [" + userId + "]");
+        }
+
+        userConsentEntity.setGrantedClientScopesIds(
+                consent.getGrantedClientScopes().stream()
+                        .map(ClientScopeModel::getId)
+                        .collect(Collectors.toSet())
+        );
+
+        userConsentEntity.setLastUpdatedDate(Time.currentTimeMillis());
```

Comment:
Not sure if I fully understand how all this works but shouldn't this update happen in a transaction?

Reply:
Well this is the opposite scenario than before. This is where we need the registered entity, otherwise it won't be done in transaction. Nice catch! I changed the method `getEntityByIdOrThrow` to `getRegisteredEntityByIdOrThrow`.